src/index.ts

```diff
@@ -267,13 +277,23 @@ export class Collection<K, V> extends Map<K, V> {
 	 * @returns {Collection[]}
 	 * @example const [big, small] = collection.partition(guild => guild.memberCount > 250);
 	 */
+	public partition<T extends V>(
+		fn: (value: V, key: K, collection: this) => value is T,
+	): [Collection<K, T>, Collection<K, Exclude<V, T>>];
 	public partition(fn: (value: V, key: K, collection: this) => boolean): [this, this];
-	public partition<T>(fn: (this: T, value: V, key: K, collection: this) => boolean, thisArg: T): [this, this];
-	public partition(fn: (value: V, key: K, collection: this) => boolean, thisArg?: unknown): [this, this] {
+	public partition<This, T extends V>(
+		fn: (this: This, value: V, key: K, collection: this) => value is T,
+		thisArg: This,
+	): [Collection<K, T>, Collection<K, Exclude<V, T>>];
+	public partition<This>(fn: (this: This, value: V, key: K, collection: this) => boolean, thisArg: This): [this, this];
+	public partition(
+		fn: (value: V, key: K, collection: this) => boolean,
+		thisArg?: unknown,
+	): [Collection<K, V>, Collection<K, V>] {
```

Comment:
If I use `[this, this]` as the return type here TypeScript will complain that the overloads returning `[Collection<K, T>, Collection<K, Exclude<V, T>>]` are not assignable to the implementation signature.